tools/state-viewer/src/state_changes.rs

```diff
@@ -0,0 +1,194 @@
+fn apply_state_changes(
+    file: PathBuf,
+    shard_id: ShardId,
+    mut state_root: StateRoot,
+    home_dir: &Path,
+    near_config: NearConfig,
+    store: Store,
+) {
+    let runtime: NightshadeRuntime =
+        NightshadeRuntime::from_config(home_dir, store.clone(), &near_config);
+    let mut chain_store = ChainStore::new(
+        store,
+        near_config.genesis.config.genesis_height,
+        near_config.client_config.save_trie_changes,
+    );
+
+    let data = std::fs::read(&file).unwrap();
+    let state_changes_for_block_range = StateChangesForBlockRange::try_from_slice(&data).unwrap();
+
+    for StateChangesForBlock { block_hash, state_changes } in state_changes_for_block_range.blocks {
+        let block_header = chain_store.get_block_header(&block_hash).unwrap();
+        let block_hash = block_header.hash();
+        let block_height = block_header.height();
+        let epoch_id = block_header.epoch_id();
+        let shard_uid = runtime.shard_id_to_uid(shard_id, epoch_id).unwrap();
+
+        for StateChangesForShard { shard_id: state_change_shard_id, state_changes } in state_changes
+        {
+            if state_change_shard_id != shard_id {
+                continue;
+            }
+
+            tracing::info!(target: "state-changes", block_height, ?block_hash, ?shard_uid, ?state_root, num_changes = state_changes.len(), "Applying state changes");
+            let trie =
+                runtime.get_trie_for_shard(shard_id, &block_hash, state_root, false).unwrap();
+
+            let trie_update = trie
+                .update(state_changes.iter().map(|raw_state_changes_with_trie_key| {
+                    tracing::debug!(target: "state-changes", ?raw_state_changes_with_trie_key);
+                    let raw_key = raw_state_changes_with_trie_key.trie_key.to_vec();
+                    let data = raw_state_changes_with_trie_key.changes.last().unwrap().data.clone();
+                    (raw_key, data)
+                }))
+                .unwrap();
+
+            tracing::info!(target: "state-change", block_height, ?block_hash, ?shard_uid, old_state_root = ?trie_update.old_root, new_state_root = ?trie_update.new_root, "Applied state changes");
+            state_root = trie_update.new_root;
+
+            let wrapped_trie_changes = WrappedTrieChanges::new(
+                runtime.get_tries(),
+                shard_uid,
+                trie_update,
+                state_changes,
+                *block_hash,
+            );
+            let mut store_update = chain_store.store_update();
+            store_update.save_trie_changes(wrapped_trie_changes);
+            store_update.commit().unwrap();
+        }
+
+        if let Ok(chunk_extra) = chain_store.get_chunk_extra(block_hash, &shard_uid) {
+            assert_eq!(chunk_extra.state_root(), &state_root);
+            tracing::debug!(target: "state-changes", block_height, ?state_root, "StateRoot Matches");
+        }
```

Comment:
optionally (in a separate PR) -
- if we have the block (that contains chunk headers) - we could compare with the next block's chunk.

Reply:
Will follow up.
ChunkHeader should be enough to get state roots. No reason to use ChunkExtra.

Reply:
Done